exchanges/poloniex/poloniex_websocket.go

```diff
@@ -28,1128 +29,507 @@ import (
+		return intervalString, kline.Interval(0), fmt.Errorf("%w %q", kline.ErrInvalidInterval, intervalString)
 	}
+	intervalValue, err := stringToInterval(strings.Join(splits[length-2:], "_"))
+	return strings.Join(splits[:length-2], "_"), intervalValue, err
+}
 
-	t, ok := notification[8].(string)
-	if !ok {
-		return fmt.Errorf("%w time not string", errTypeAssertionFailure)
-	}
-	timeParse, err := time.Parse(time.DateTime, t)
+// Subscribe sends a websocket message to receive data from the channel
+func (e *Exchange) Subscribe(ctx context.Context, conn websocket.Connection, subs subscription.List) error {
+	subs, err := subs.ExpandTemplates(e)
 	if err != nil {
 		return err
 	}
+	return e.ParallelChanOp(ctx, subs, func(ctx context.Context, l subscription.List) error {
+		return e.manageSubs(ctx, conn, "subscribe", l)
+	}, 1)
+}
 
-	// null returned so ok check is not needed
-	clientOrderID, _ := notification[9].(string)
-
-	tt, ok := notification[10].(string)
-	if !ok {
-		return fmt.Errorf("%w time not string", errTypeAssertionFailure)
-	}
-	tradeTotal, err := strconv.ParseFloat(tt, 64)
+// Unsubscribe sends a websocket message to stop receiving data from the channel
+func (e *Exchange) Unsubscribe(ctx context.Context, conn websocket.Connection, subs subscription.List) error {
+	subs, err := subs.ExpandTemplates(e)
 	if err != nil {
 		return err
 	}
-
-	e.Websocket.DataHandler <- &order.Detail{
-		Exchange: e.Name,
-		OrderID:  strconv.FormatFloat(orderID, 'f', -1, 64),
-		Fee:      totalFee,
-		Trades: []order.TradeHistory{{
-			Price:     rate,
-			Amount:    amount,
-			Fee:       totalFee,
-			Exchange:  e.Name,
-			TID:       strconv.FormatFloat(tradeID, 'f', -1, 64),
-			Timestamp: timeParse,
-			Total:     tradeTotal,
-		}},
-		AssetType:     asset.Spot,
-		ClientOrderID: clientOrderID,
-	}
-	return nil
+	return e.ParallelChanOp(ctx, subs, func(ctx context.Context, l subscription.List) error {
+		return e.manageSubs(ctx, conn, "unsubscribe", l)
+	}, 1)
 }
 
-func (e *Exchange) processAccountKilledOrder(notification []any) error {
-	if len(notification) < 3 {
-		return errNotEnoughData
-	}
-
-	orderID, ok := notification[1].(float64)
-	if !ok {
-		return fmt.Errorf("%w order ID not float64", errTypeAssertionFailure)
-	}
-
-	// null returned so ok check is not needed
-	clientOrderID, _ := notification[2].(string)
-
-	e.Websocket.DataHandler <- &order.Detail{
-		Exchange:      e.Name,
-		OrderID:       strconv.FormatFloat(orderID, 'f', -1, 64),
-		Status:        order.Cancelled,
-		AssetType:     asset.Spot,
-		ClientOrderID: clientOrderID,
+func (e *Exchange) manageSubs(ctx context.Context, conn websocket.Connection, operation string, subs subscription.List) error {
+	var errs error
+	for _, s := range subs {
+		if strings.HasSuffix(conn.GetURL(), "private") != s.Authenticated {
+			continue
+		}
+		payload, err := e.handleSubscription(operation, s)
+		if err != nil {
+			errs = common.AppendError(errs, err)
+			continue
+		}
```

Comment:
🚧 Change request:
I was asking for a Lock here, because of the error handling.
What you're doing instead is forcing a long wait before not tying up with the error.
Now we've come this far, please use an instance field lock.
I know it sucks to make them go sequentially, but I don't see a choice.

Reply:
But if we are subscribing to one at a time, no need to use lock at all ... because `SendMessageReturnResponse` is a blocking method call.

Reply:
Sorry, but this makes no sense at all.
Creating a waitgroup, to manage goroutines... which then lock and single step....
is exactly the same as running them sequentially without a lock and without a goroutine.

I think that maybe the misunderstanding here is about the assumption that manageSubs wouldn't be called concurrently. I don't thing we can guarantee that at all.
We don't control the execution routes to `Subscribe`. Users are free to call it concurrently.